firestore/integration_test_internal/src/aggregate_query_snapshot_test.cc

```diff
@@ -46,90 +46,339 @@ std::size_t AggregateQuerySnapshotHash(const AggregateQuerySnapshot& snapshot) {
+
+  EXPECT_EQ(snapshot_copy_dest.count(), 7);
+  EXPECT_EQ(snapshot_copy_dest.query(), aggregate_query);
+  EXPECT_TRUE(snapshot_copy_dest.is_valid());
+
+  snapshot_copy_dest = snapshot;
+
+  EXPECT_EQ(snapshot.count(), 0);
+  EXPECT_EQ(snapshot.query(), AggregateQuery());
+  EXPECT_FALSE(snapshot.is_valid());
+
+  EXPECT_EQ(snapshot_copy_dest.count(), 0);
+  EXPECT_EQ(snapshot_copy_dest.query(), AggregateQuery());
+  EXPECT_FALSE(snapshot_copy_dest.is_valid());
+}
+
+TEST_F(
+    AggregateQuerySnapshotTest,
+    ValidObjectCopyAssignmentOperatorAppliedToDefaultObjectReturnsEqualObject) {
+  const AggregateQuery aggregate_query1 =
+      TestFirestore()->Collection("foo").Limit(10).Count();
+  const AggregateQuery aggregate_query2 =
+      TestFirestore()->Collection("bar").Limit(20).Count();
+
+  const AggregateQuerySnapshot snapshot =
+      TestAggregateQuerySnapshot(aggregate_query1, 1);
+
+  EXPECT_EQ(snapshot.count(), 1);
+  EXPECT_EQ(snapshot.query(), aggregate_query1);
+  EXPECT_TRUE(snapshot.is_valid());
+
+  AggregateQuerySnapshot snapshot_copy_dest =
+      TestAggregateQuerySnapshot(aggregate_query2, 2);
+
+  EXPECT_EQ(snapshot_copy_dest.count(), 2);
+  EXPECT_EQ(snapshot_copy_dest.query(), aggregate_query2);
+  EXPECT_TRUE(snapshot_copy_dest.is_valid());
+
+  snapshot_copy_dest = snapshot;
+
+  EXPECT_EQ(snapshot.count(), 1);
+  EXPECT_EQ(snapshot.query(), aggregate_query1);
+  EXPECT_TRUE(snapshot.is_valid());
+
+  EXPECT_EQ(snapshot_copy_dest.count(), 1);
+  EXPECT_EQ(snapshot_copy_dest.query(), aggregate_query1);
+  EXPECT_TRUE(snapshot_copy_dest.is_valid());
+}
+
+TEST_F(AggregateQuerySnapshotTest,
+       CopyAssignmentAppliedSelfReturnsEqualObject) {
+  const AggregateQuery aggregate_query =
+      TestFirestore()->Collection("foo").Limit(10).Count();
 
-  static const int COUNT = 8;
   AggregateQuerySnapshot snapshot =
-      TestAggregateQuerySnapshot(aggregate_query, COUNT);
+      TestAggregateQuerySnapshot(aggregate_query, 7);
+
+  EXPECT_EQ(snapshot.count(), 7);
+  EXPECT_EQ(snapshot.query(), aggregate_query);
+  EXPECT_TRUE(snapshot.is_valid());
+
+  snapshot = snapshot;
+
+  EXPECT_EQ(snapshot.count(), 7);
+  EXPECT_EQ(snapshot.query(), aggregate_query);
+  EXPECT_TRUE(snapshot.is_valid());
+}
+
+TEST_F(AggregateQuerySnapshotTest,
+       MoveConstructorAppliedToValidObjectReturnsEqualObject) {
+  const AggregateQuery aggregate_query =
+      TestFirestore()->Collection("foo").Limit(10).Count();
+
+  AggregateQuerySnapshot snapshot =
+      TestAggregateQuerySnapshot(aggregate_query, 11);
+
+  EXPECT_EQ(snapshot.count(), 11);
+  EXPECT_EQ(snapshot.query(), aggregate_query);
+  EXPECT_TRUE(snapshot.is_valid());
 
   AggregateQuerySnapshot moved_snapshot_dest(std::move(snapshot));
 
   EXPECT_EQ(snapshot.count(), 0);
   EXPECT_EQ(snapshot.query(), AggregateQuery());
   EXPECT_FALSE(snapshot.is_valid());
 
-  EXPECT_EQ(moved_snapshot_dest.count(), COUNT);
+  EXPECT_EQ(moved_snapshot_dest.count(), 11);
```

Comment:
When testing move constructors and move assignment operators, don't perform assertions on the "moved-from" object after it has been moved from. The C++ standard only requires that moved-from objects support three methods: the destructor, the copy assignment operator, and the move assignment operator. Any other method calls on a moved-from object are undefined. Of course, classes are free to make stronger guarantees about the state of a moved-from object, but doing so is generally ill advised. By adding tests for the state of a moved-from object you are implicitly making this stronger guarantee which has minimal/no value. So it's best to just leave a moved-from object alone, even in the unit tests.

Reply:
Hmm that's a good point. Ok fair enough. I'm still torn because calling methods on a moved-from object is technically undefined but you make a sounds argument to test for it nontheless.